[FIXED JENKINS-37590] Better handling of null parameter definitions.

abayer opened the pull request:
[JENKINS-37590](https://issues.jenkins-ci.org/browse/JENKINS-37590)

Rejecting null parameter definitions lists results in an error if you
try to save a job config with "This build is parameterized" checked
but no actual parameters specified. We should handle null better by
populating an empty list if needed instead.

cc @jenkinsci/code-reviewers @reviewbybees

abayer committed:
[FIXED JENKINS-37590] Better handling of null parameter definitions.

Rejecting null parameter definitions lists results in an error if you
try to save a job config with "This build is parameterized" checked
but no actual parameters specified. We should handle null better by
populating an empty list if needed instead.

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

imod commented:
👍

oleg-nenashev commented on core/src/main/java/hudson/model/ParametersDefinitionProperty.java:
```diff
@@ -70,18 +70,19 @@
 
     @DataBoundConstructor
-    public ParametersDefinitionProperty(@Nonnull List<ParameterDefinition> parameterDefinitions) {
+    public ParametersDefinitionProperty(List<ParameterDefinition> parameterDefinitions) {
```
Add CheckForNull instead

oleg-nenashev commented on core/src/main/java/hudson/model/ParametersDefinitionProperty.java:
```diff
@@ -70,18 +70,19 @@
-    public ParametersDefinitionProperty(@Nonnull ParameterDefinition... parameterDefinitions) {
-        if (parameterDefinitions == null) {
-            throw new NullPointerException("ParameterDefinitions is null when this is a not valid value");
+    public ParametersDefinitionProperty(ParameterDefinition... parameterDefinitions) {
```
Also CheckForNull

oleg-nenashev reviewed (approved):
I would rather like to see one constructor invoking another one && Annotations && Javadoc. But 🐝

jglick commented:
IIUC corrects some regression in #2296?

jglick reviewed (changes requested):
:-1:, would rather just fix [JENKINS-33991](https://issues.jenkins-ci.org/browse/JENKINS-33991) so that there is no property at all in this case.

oleg-nenashev commented:
Maybe needs-fix since @jglick blocks it. But fine for me

daniel-beck commented:
Depends which behavior makes more sense on the UI. "Build With Parameters" with an empty page or something similar that would appear broken?

oleg-nenashev commented:
> Depends which behavior makes more sense on the UI. "Build With Parameters" with an empty page or something similar that would appear broken?

Well, it's rather a Unix way. If one feel brave, he can do it (e.g. if one injects Javascript magic into this Build With Parameters page). Empty screen by default is not that bad.

oleg-nenashev commented:
Of course a fallback text like "No parameters defined" would be even better

daniel-beck commented:
> If one feel brave, he can do it (e.g. if one injects Javascript magic into this Build With Parameters page). Empty screen by default is not that bad.

We shouldn't elevate injection of JS/CSS into a supported use case. There's just pain down that road. So this isn't an argument for me.

oleg-nenashev commented:
@daniel-beck So would you be happy if there is a "No parameters defined" in the "buildParameterized" page?

daniel-beck commented:
@oleg-nenashev 

>  So would you be happy if there is a "No parameters defined" in the "buildParameterized" page?

No, I think the job should not be considered to be parameterized if there are no parameters.

oleg-nenashev commented:
OK, then needs fix

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

abayer committed:
Add CheckForNulls

abayer committed:
Add test verifying that JENKINS-37590 now results in correct error

abayer commented:
So I don't get how [JENKINS-33991](https://issues.jenkins-ci.org/browse/JENKINS-33991) should be addressed. With https://github.com/jenkinsci/jenkins/commit/0b517705f5de454f2896d5c345ac8ec4fa8f6651 in from 2.14 onward, it's no longer possible to save a job with empty parameters and the parameterized button checked, though the fix here was still needed to deal with some edge cases. Anyway, I've updated responding to review comments and added a round-tripping test that gets the right result.

jglick commented:
IMO better to revert #2444, and adopt my suggestion from JENKINS-33991 to restore the historical behavior in Jenkins: override `ParametersDefinitionProperty.DescriptorImpl.newInstance` to return `null` in case the `super` return value had `parameterDefinitions.isEmpty()`. Easy, no UI hassle.

jglick reviewed (changes requested):
As above.

abayer commented:
@jglick #2444 was already reverted in https://github.com/jenkinsci/jenkins/commit/e43222dde84be0ea7d05647790fedc12d70f8052#diff-49180eb94a7d79905023758a89bb88eb after I opened this...so I'm just junking this PR completely and starting over. Yay?
